### PR TITLE
[BEAM-4775] mv function pkg from fn-harness to java-core

### PR DESCRIPTION
**Pull request by ryan-williams**
also: parameterize ThrowingConsumer by Exception type

This is factored out of #7823, where a use-case for a `ThrowingConsumer<IOException, …>` exists, in Jackson JSON serde code.

R: @robertwb, @mxm 

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

See [.test-infra/jenkins/README](../.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ryan-williams** [comment]
Yea, sorry it's not clear. I need to think more about metrics/proto package-structures in #7823, that may affect this. Let me do that and circle back. 

In the meantime, here's some context for the package-move, as it currently stands:

- in #7823, [I have a use for `ThrowingConsumer` in `sdks/java/core`](https://github.com/apache/beam/pull/7823/files#diff-e819ee62f70ca51640a1d798c8a9befeR72)
- it (and its friends) currently live in `sdks/java/fn-harness`
  - they don't have anything to do with the fn-harness
  - ideally, they'd be in `java.util.function`, or something like Guava; they don't have anything to do with Beam
- `sdks/java/core` seems the most logical place to put them that is accessible from `sdks/java/core`
  - I could also make a new module, and have java-core depend on that
    - However, that has all the downsides from a "lots of things now transitively depend on this" perpspective
    - It would keep things a bit more modular in general though, and leave the option to rearrange things later, better parallelize builds in the meantime, etc., so maybe that's worth doing.
  - It's also possible I can move the java-core metrics stuff to a different, less "core", package
    - Let me think about this for a bit, I need to think more about where all the Java SDK metrics stuff should live in #7823 anyway.

re: parameterizing ThrowingConsumer`'s `Exception`-type:
- I pass a JSON-serialization action as a lambda and need to encode that it `throws IOException`
- If [the code that receives that lambda](https://github.com/apache/beam/pull/7823/files#diff-e819ee62f70ca51640a1d798c8a9befeR72) declares a generic `throws Exception`, then [code that calls that](https://github.com/apache/beam/pull/7823/files#diff-1677e5001d798d26293b99815a5c16d8R114) would have to do some unwieldy catching/casting to go from `Exception` back to `IOException`.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/function/ThrowingConsumer.java [hunk 1]
```diff
@@ -26,6 +26,6 @@
  * interfaces.
  */
 @FunctionalInterface
-public interface ThrowingConsumer<T> {
-  void accept(T t) throws Exception;
+public interface ThrowingConsumer<ExceptionT extends Exception, T> {
+  void accept(T t) throws ExceptionT;
```
Comment (mxm) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/function/ThrowingConsumer.java]: What is the use case for this change? It looks like you only pass `Exception` as `ExceptionT`. Do you want to define contracts for other subclasses of `Exception`?

Comment (ryan-williams) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/function/ThrowingConsumer.java]: There was a note at the end of [this comment](https://github.com/apache/beam/pull/7866#issuecomment-465241925) about that part.

It's not used here, but I do have [a motivating case for a `ThrowingConsumer<IOException, T>`](https://github.com/apache/beam/pull/7823/files#diff-1677e5001d798d26293b99815a5c16d8R114) in #7823.

**Comment by ryan-williams** [comment]
I think this can be decoupled from metrics/module concerns, and is good to go, if you agree that java/core is a more logical place for such generic language-level helpers.

Even if the metrics code that uses it ends up somewhere higher-up than java/core, having it depend on these helpers in java/core seems reasonable.

Pulling things like this back out of java/core to dedicated packages later seems fine and good as well, if necessary.

**Comment by mxm** [comment]
Looks good. Thanks!

**Pull request merged**